File: leetcode/add-binary/add-binary.java
/**
 * Title: add-binary
 * Slug: add-binary
 * URL: https://leetcode.com/problems/add-binary/?envType=problem-list-v2&envId=string
 * Language: java
 * Submitted: 2025-10-21T18:24:19.507Z
 */

class Solution {
    public String addBinary(String a, String b) {
        int i = a.length() - 1, j = b.length() - 1, carry = 0;
        StringBuilder sb = new StringBuilder();

        while (i >= 0 || j >= 0 || carry != 0) {
            int sum = carry;
            if (i >= 0) sum += a.charAt(i--) - '0';
            if (j >= 0) sum += b.charAt(j--) - '0';
            sb.append(sum % 2);
            carry = sum / 2;
        }

        return sb.reverse().toString();
    }
}

File: leetcode/add-binary/add-binary.py
/**
 * Title: add-binary
 * Slug: add-binary
 * URL: https://leetcode.com/problems/add-binary/?envType=problem-list-v2&envId=string
 * Language: python
 * Submitted: 2025-10-21T18:23:54.058Z
 */

class Solution(object):
    def addBinary(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: str
        """
        i, j = len(a) - 1, len(b) - 1
        carry = 0
        res = []

        while i >= 0 or j >= 0 or carry:
            total = carry
            if i >= 0:
                total += int(a[i])
                i -= 1
            if j >= 0:
                total += int(b[j])
                j -= 1
            res.append(str(total % 2))
            carry = total // 2

        return ''.join(reversed(res))
